### PR TITLE
fix: add support for negative values via keyboard

### LINKED ISSUE #2492: Power Souce: EditTexts don't allow negative values for PV1 and PV2
**Actual Behaviour**

The values for PV1 and PV2 may be negative, but the `EditText`s for PV1 and PV2 don't allow to enter negative values.

**Expected Behaviour**

The `EditText`s should allow negative values for PV1 and PV2.

It is possible to allow negative decimal numbers as values by setting the `android:inputType` to `numberSigned|numberDecimal`, but this introduces problems due to a bug in the keyboards of older Android version (before Android O). A workaround is required. See https://stackoverflow.com/questions/3821539/decimal-separator-comma-with-numberdecimal-inputtype-in-edittext/51983392 for more information and hints.

**Steps to reproduce it**

1. Open app
2. Open Power Souce
3. Try to enter negative value for PV1 or PV2 with keyboard

**Would you like to work on the issue?**
Maybe

### PR DESCRIPTION
**Pull request by Dishebh**
Fixes #2492 

## Changes 
- Added support for entering negative values via mobile keyboard
- Added support for entering comma (",") instead of decimal (".") for different locales

## Screenshots / Recordings  

**Checklist**: 
- [ ] **No hard coding**: I have used resources from `strings.xml`, `dimens.xml` and `colors.xml` without hard coding any value.
- [ ] **No end of file edits**: No modifications done at end of resource files `strings.xml`, `dimens.xml` or `colors.xml`.
- [ ] **Code reformatting**: I have reformatted code and fixed indentation in every file included in this pull request.
- [ ] **No extra space**: My code does not contain any extra lines or extra spaces than the ones that are necessary.

### CODE REVIEW

**Review by AsCress — APPROVED** [review]
@Dishebh Thanks very much for your work on this issue ! This works perfectly.
Your thoughts on this @marcnause ?

**Comment by Dishebh** [comment]
Thanks @AsCress ! 

On a separate note - whenever I make a change, even as small as changing the inputType, I have to rebuild the entire gradle to see the change reflect in the App emulator. Is there any quick way you would know of, kind of like hot module reloading, which I can apply here? I have googled this similar query, but I haven't been able to see it in action in my android studio (maybe they removed it in latest versions, or could I missing a step?). TIA!

**Comment by AsCress** [comment]
@Dishebh There's no feature like the _hot reload_ (Flutter) in native development unfortunately. 
However, once you build an app, gradle shouldn't take much time for a re-build whenever you make any small change in there, if the gradle _caches_ are in place and the already started gradle _daemon_ is being re-used (which is usually the case by default).
On my system, it only takes something like 5 seconds for a re-build whenever I make any change (which quite matches the _hot reload_ time in Flutter).

**Comment by Dishebh** [comment]
Got it, thank you @AsCress !

@marcnause can you please review it when you get some time?

**Comment by AsCress** [comment]
@Dishebh No issues ! 
Just a simple suggestion, if you want reviews from anyone, you could request them using the option at the top right in every PR :))
Happy coding!

**Comment by Dishebh** [comment]
Screenshot 2024-09-17 at 11 27 26 PM

I actually tried, but somehow, I don't see an option to request review from anyone 😅 . I think there are additional permissions for this, which I may not have lol.

File: app/src/main/java/io/pslab/activity/PowerSourceActivity.java [hunk 1]
```diff
@@ -214,7 +214,8 @@ public void onClick(View v) {
             public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
                 if (actionId == EditorInfo.IME_ACTION_DONE) {
                     final String voltageValue = remove(displayPV1.getText(), "V", "\\+").trim();
-                    final float voltage = PV1_VOLTAGE_RANGE.clamp(parseFloat(voltageValue, PV1_VOLTAGE_RANGE.getLower()));
+                    final String decimalVoltageValue = voltageValue.replace(",", ".");
+                    final float voltage = PV1_VOLTAGE_RANGE.clamp(parseFloat(decimalVoltageValue, PV1_VOLTAGE_RANGE.getLower()));
```
[the remark this answers is not in the document]
Comment (marcnause) [reply app/src/main/java/io/pslab/activity/PowerSourceActivity.java:217]: It should not be possible to enter illegal values thanks to the limited range of values defined for the corresponding TextView.

**Comment by marcnause** [comment]
I had a problem when I tried the app on my device (German locale). Parsing failed since in German locale a comma is used instead of the decimal point. Since all commas are replaced with points, I got an exception (and consequently the default value was used as a fallback).

I changed the locale which is assumed for parsing (German on my device) to `Locale.ROOT` which makes sure that it does not matter which locale is used on the device.

**Review by marcnause — APPROVED** [review]

**Pull request merged**